lib/elixir/lib/calendar/iso.ex

```diff
@@ -280,50 +357,92 @@ defmodule Calendar.ISO do
   def parse_naive_datetime(string) when is_binary(string),
     do: parse_naive_datetime(string, 1)
 
-  defp parse_naive_datetime(string, multiplier) do
-    with <<unquote(match_date), sep, unquote(match_time), rest::binary>> <- string,
-         true <- unquote(guard_date) and sep in @sep and unquote(guard_time),
-         {microsecond, rest} <- parse_microsecond(rest),
+  defp parse_naive_datetime(
+         <<unquote(match_calendar_date), sep, unquote(match_time), rest::binary>>,
+         multiplier
+       )
+       when unquote(guard_calendar_date) and sep in @sep and unquote(guard_time) do
+    with {microsecond, rest} <- parse_microsecond(rest),
          {_offset, ""} <- parse_offset(rest) do
-      {year, month, day} = unquote(read_date)
+      {year, month, day} = unquote(read_calendar_date)
       {hour, minute, second} = unquote(read_time)
       year = multiplier * year
 
-      cond do
-        not valid_date?(year, month, day) ->
-          {:error, :invalid_date}
+      validate_parsed_naive_datetime(year, month, day, hour, minute, second, microsecond)
+    else
+      _ -> {:error, :invalid_format}
+    end
+  end
 
-        not valid_time?(hour, minute, second, microsecond) ->
-          {:error, :invalid_time}
+  defp parse_naive_datetime(
+         <<unquote(match_ordinal_date), sep, unquote(match_time), rest::binary>>,
+         multiplier
+       )
+       when unquote(guard_ordinal_date) and sep in @sep and unquote(guard_time) do
+    with {microsecond, rest} <- parse_microsecond(rest),
+         {_offset, ""} <- parse_offset(rest) do
+      {year, ordinal_day} = unquote(read_ordinal_date)
+
```

Comment:
Worth noting: the implementation of `calendar_day_of_year` returns the year as well, and could trivially "roll over" the year if given an ordinal day beyond its range (my original impl did this). However, this is against the ISO-8601 spec, so we handle it here instead, and similarly in the other parsing functions.